src/gil.rs

```diff
@@ -425,7 +449,14 @@ pub unsafe fn register_owned(_py: Python<'_>, obj: NonNull<ffi::PyObject>) {
 #[inline(always)]
 fn increment_gil_count() {
     // Ignores the error in case this function called from `atexit`.
-    let _ = GIL_COUNT.try_with(|c| c.set(c.get() + 1));
+    let _ = GIL_COUNT.try_with(|c| {
+        let current = c.get();
+        assert!(
+            current >= 0,
+            "Access to the GIL is currently prohibited, for example because a `__traverse__` implementation is currently running."
+        );
```

Comment:
We could potentially even use different negative integers to emit different error messages, if we wanted to be more definite here that a `__traverse__` _is_ running.

Reply:
Will add a named constant and a definite assertion message.

Reply:
Add the constant and assertion message. It does look a bit more complicated now and out of an abundance of caution, I moved choosing the correct panic message into a cold function.